[WebGPU] Packed struct types have unexpected sizes

mwyrzykowski opened the pull request:
#### 9dad9cc100737bd8ab26e356fff29b1299ba2e28
<pre>
[WebGPU] Packed struct types have unexpected sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281933">https://bugs.webkit.org/show_bug.cgi?id=281933</a>
<a href="https://rdar.apple.com/138380467">rdar://138380467</a>

Reviewed by Dan Glastonbury.

The specification says that the alignment must be a multiple
of the required alignment.

* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::validateAlignment):

Canonical link: <a href="https://commits.webkit.org/285722@main">https://commits.webkit.org/285722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6140254b59f486c471aff5b13e0d19124f44ddf6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24827 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57853 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16261 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48001 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38259 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44593 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23160 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66339 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21150 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79476 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/384 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66225 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1048 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63325 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65506 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9359 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7537 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/870 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/899 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/886 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/905 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/AttributeValidator.cpp:
```diff
@@ -305,7 +305,8 @@ void AttributeValidator::validateAlignment(const SourceSpan& span, AddressSpace
-            if (member.offset() % requiredAlignment(type))
+            auto alignment = requiredAlignment(type);
+            if ((member.offset() % alignment) || (member.alignment() % alignment))
                 error(member.span(), "offset of struct member "_s, structure.name(), "::"_s, member.name(), " must be a multiple of "_s, String::number(requiredAlignment(type)), " bytes, but its offset is "_s, String::number(member.offset()), " bytes"_s);
```
Does this message need to be updated to say, "offset and alignment of struct member ...".

Even better would be to separate the two tests for `member.offset()` and `member.alignment()` to give the specific error message, "offset of ..." or "alignment of ..."

mwyrzykowski replied:
Made two separate messages, thanks for the suggestion 👍

djg reviewed (approved):
🐞

mwyrzykowski committed:
[WebGPU] Packed struct types have unexpected sizes
https://bugs.webkit.org/show_bug.cgi?id=281933
rdar://138380467

Reviewed by Dan Glastonbury.

The specification says that the alignment must be a multiple
of the required alignment.

* Source/WebGPU/WGSL/AttributeValidator.cpp:
(WGSL::AttributeValidator::validateAlignment):

Canonical link: https://commits.webkit.org/285722@main
